### PR TITLE
dropdown editor implementation

### PR DESCRIPTION
**Pull request by titovmx**
- add editor options to column model
- add dropdown editor template
- add fetch class
- add demo page for editors

### CODE REVIEW

File: demo/exception.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ export default function Handler($log, $injector) {
 		return toast;
 	}
 
-	return (exception, cause) => {
+	return (exception, cause='qGrid') => {
```
Comment (klumba12) [inline demo/exception.js]: renmae it to 'qgrid' please :)

File: demo/pages/editors/index.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+			path: 'name.first'
+		},
+		{
+			key: 'gender',
+			title: 'Is Female',
+			type: 'bool',
+			value: (item, value) => isUndef(value) ? item.gender === 'female' : item.gender = value,
+			editor: 'dropdown',
+			editorOptions: {
+				fetch: ['female', 'male']
```
Comment (klumba12) [inline demo/pages/editors/index.js:29]: can you add an example where d.resolve is used?

File: src/core/column-type/column.model.js [hunk 3]
```diff
@@ -7,6 +7,10 @@ export default class ColumnModel {
 		this.pin = null;
 		this.origin = 'specific';
 		this.source = 'user';
+		this.editor = null;
+		this.editorOptions = {
+			editTrigger: 'click'
+		};
```
Comment (klumba12) [inline src/core/column-type/column.model.js]: rename it to trigger if it's insed editor-options

File: src/core/column-type/email.column.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ class EmailColumnModel extends DataColumnModel {
 	constructor() {
 		super('email');
 
-		this.canEdit = false;
+		this.editorOptions.editTrigger = 'button';
```
Comment (klumba12) [inline src/core/column-type/email.column.js]: i think 'custom' will be better and default name? Button knows about template too much..

File: src/core/edit/edit.cell.view.js [hunk 5]
```diff
@@ -28,6 +29,8 @@ export default class EditCellView {
 		this.commit = commands.get('commit');
 		this.cancel = commands.get('cancel');
 		this.reset = commands.get('reset');
+		
+		this.fetch = null;
```
Comment (klumba12) [inline src/core/edit/edit.cell.view.js]: can we provide default fetch that will return edit cell value?

File: src/core/infrastructure/fetch.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+export default class Fetch {
+	constructor(source) {
+		this.source = source;
+		this.busy = null;
+		this.result = null;
+	}
+
+	run() {
+		const fetch = isFunction(this.source) ? this.source(...arguments) : this.source;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.js]: const p = new Promise()
const fecth = this.source(...arguments, {resolve: () => Promise,.resolve(p), reject: Promise.reject(p});

File: demo/pages/editors/index.js (hunk 2 above)
Comment (klumba12) [reply demo/pages/editors/index.js:29]:
```javascript
fetch: function(item, d){
d.resolve(['female', 'male'])
}

fetch: function(item, d){
return ['female', 'male'];
}

fetch: function(item, d){
  return $q.promise
}
fetch: ['foo', 'bar']


```

File: src/ng/components/body/td.core.js [hunk 7]
```diff
@@ -60,6 +60,9 @@ class TdCore extends Directive(TD_CORE_NAME, {view: `^^${VIEW_CORE_NAME}`}) {
 				}
 
 				link(this.$element, templateScope);
+				if (column.editor === 'dropdown') {
```
Comment (klumba12) [inline src/ng/components/body/td.core.js]: get rid of that

Comment (klumba12) [reply src/ng/components/body/td.core.js]: try ng-init $cell.$element.click()

File: src/core/infrastructure/fetch.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import {isFunction, noop} from 'core/services/utility';
+
+export default class Fetch {
+	constructor(source) {
```
Comment (klumba12) [inline src/core/infrastructure/fetch.js]: rename to select

File: src/core/infrastructure/fetch.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+		const fetch = isFunction(this.source) ? this.source(...arguments) : this.source;
+		this.busy = new Promise(resolve => {
+			Promise.resolve(fetch).then(result => {
+				this.result = result;
+				if (this.busy) {
+					resolve(false);
+				}
+			}).catch(ex => {
+				resolve(false);
+				throw ex;
```
Comment (vkorolev) [inline src/core/infrastructure/fetch.js]: `reject(ex)` maybe?

**Review by klumba12 — APPROVED** [review]

**Pull request merged**